sql/parser: improved utf8 rune error handling

madelynnblue opened the pull request:
There is some input that causes Go's string range rune to produce a
RuneError that has length != 1, as would be expected for a RuneError
due to the Go spec about range on strings. This input is still valid
(utf8.ValidString is true). Allow encodeSQLString to correctly output
the bytes for these cases. Note that we cannot simply use the normal
paths below the RuneError path because r == '\uFFFD', but in[i:i+ln]
!= '\uFFFD'.

Fixes #12156

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12194)
<!-- Reviewable:end -->

knz commented:
I think I'm ok with the change but your commit message needs more words about how it solves #12156:

1) I think your code is now detecting more errors on input. Is that good enough? What do we do with cases where invalid strings are already in the database? If we don't want to do anything about that (like the last time I looked at this part of the code) then at least we should indicate so in the commit message and later the release notes.

2) can you give an example in the commit message that failed before and doesn't any more (perhaps copy paste from Sean's original message)

mjibson commented:
The new TestEncodeSQLStringSpecial test now succeeds where it would have failed before. This isn't detecting errors on input, it's attempting to correctly encode data for output to, i.e., a dump. The problem is that a RuneError can sometimes be the actual input, in which case it doesn't have the normal byte length of 1, but instead 3.

knz commented:
Ok thanks for explaining.

Can you add some of these words to the commit message as well? This will help if we need to understand this code later.

mjibson commented:
I've added a better description to the commit message.

knz commented:
Excellent :+1:

LGTM

madelynnblue committed:
sql/parser: improved utf8 rune error handling

There is some input that causes Go's string range rune to produce a
RuneError that has length != 1, as would be expected for a RuneError
due to the Go spec about range on strings. This input is still valid
(utf8.ValidString is true). Allow encodeSQLString to correctly output
the bytes for these cases. Note that we cannot simply use the normal
paths below the RuneError path because r == '\uFFFD', but in[i:i+ln]
!= '\uFFFD'.

Fixes #12156
